File: Minimum Cost to cut a board into squares.cpp
class Solution {
public:
    int minimumCostOfBreaking(vector<int> X, vector<int> Y, int M, int N){
        //Write your code here
    
        int ans = 0;
        
        sort(X.begin(), X.end(), greater<int>());
        sort(Y.begin(), Y.end(), greater<int>());
        
        int i=0, j=0;
        
        while(i<M-1 and j<N-1)
        {
            if(X[i] >= Y[j])
            ans += (j+1)*X[i++];
            else
            ans += (i+1)*Y[j++];
        }
        
        while(i<M-1)
        ans += (j+1)*X[i++];
        
        while(j<N-1)
        ans += (i+1)*Y[j++];
        
        return ans;
    }
};

File: Minimum Cost to cut a board into squares.java
class Solution {
    public static int minimumCostOfBreaking(int[] X, int[] Y,int M,int N) {
        // code here
        Arrays.sort(X);
        Arrays.sort(Y);
        int i=X.length-1;
        int j=Y.length-1;
        int row=1;
        int col=1;
        int ans=0;
        while(i>=0 || j>=0){
            int a=i>=0?X[i]:-1;
            int b=j>=0?Y[j]:-1;
            if(a>=b){
                ans+=(a*row);
                col++;
                i--;
            }else{
                ans+=(b*col);
                row++;
                j--;
            }
        }
        return ans;
    }
}
        
